platform/linux-generic/odp_crypto.c

```diff
@@ -137,15 +139,56 @@ struct odp_crypto_global_s {
 	odp_spinlock_t                lock;
 	odp_crypto_generic_session_t *free;
 	odp_crypto_generic_session_t  sessions[MAX_SESSIONS];
+
+	/* These bitfields are cleared at alloc_session()
+	 * together with the rest of data */
+	bitset_t hmac_valid[ODP_THREAD_COUNT_MAX]
```

Comment:
I think the benefits of using a bitset may not outweigh the drawbacks. In addition to making the code slightly more complicated, using a bitset forces us to use atomic operations (since, as you pointed out, we still have to handle the case where thread A allocates session X while thread B is initializing session Y in the same bit set, both updating the same memory location, although different bits).

By using a full byte for the validity flag would simplify the code and avoid the need for atomics since different threads would no longer write and read the same memory location at the same time (there must be some synchronization in the application side to ensure that a session is not destroyed when still in use and that creation of a session happens before its use).

The atomics can slow down ARM because the bitset tries to provide atomicity over a 16 byte range, which causes additional overhead (e.g. reading a valid bit is not just a plain memory load) even though not required in this context. A weaker atomic bitset would suffice, but I think we could do without atomics altogether.

A benefit of the bit set is reduced memory use (and possibly some improvement in memory access locality), but I doubt it is significant enough. The savings are less than MAX_SESSIONS * ODP_THREAD_COUNT_MAX bytes (after combining the crypto/auth valid bits as in the last path). At the same time the thread local crypto_local_t structs are using 12 or 24 times MAX_SESSIONS bytes per thread, depending on pointer size, plus the memory reserved by OpenSSL for all the pre-created contexts.

So, how about just using a full byte for the validity flag?

Reply:
I feel uneasy for propagation of clear flags to other threads. What about using `atomic_flag` and `atomic_flag_test_and_set`/`atomic_flag_clear` from stdatomics?

Reply:
After some more pondering I still think additional synchronization for the valid flag is not necessary.

When an application creates a crypto session and gets a handle back, the side effects (memory stores) of the creation are not yet necessarily visible in other threads (according to Petri, this is an intended property of the ODP API). If the application passes the handle to another thread it must use some synchronization mechanism to ensure that creation is fully visible in the other thread before it tries to use the handle. The valid flag is just one more item in shared memory that needs to be visible before a crypto operation is started. Right?

Similarly, when a crypto session is destroyed, the crypto operations must have fully completed before session destroy can be called in another thread. This again requires some synchronization by the application between the threads, but that also ensures that setting of the valid bit inside a crypto operation happens-before session destroy which happens-before (due to the locking in session alloc and free) the storage of the session is reused which happens-before clearing the valid bit. So this case should also be ok, right?

And if the valid flags are all in their own memory locations (i.e. not bitfields of the same location), then there are no other inter-thread synchronization concerns regarding the valid flag.

But this requires more than just my word. If you cannot get convinced about not needing additional synchronization for the valid flag, then you could also consider memory barriers after each write to the valid flag instead.

If you want to stick with atomic ops, then I am not sure if using stdatomic is ok as it seems to create an additional dependency to a tool chain that supports it. Maybe ODP's own atomic primitives should be used.

I do not see why test-and-set would be needed.

Reply:
@JannePeltonen this sounds logical, but not 101% convincing, sorry. When we are talking about security, I'll try to be on a safe side. Maybe I have to give it another thought.

The rest of linux-generic code heavily uses atomic operation (even worse than stdatomic, it uses GCC builtins). We do not have (even internal) API for memory barriers, so I assumed that atomic flags are best fit for this case (and each flags uses just one byte).

Reply:
Is there a particular sequence that worries you or is it just general uneasiness?

Maybe you should spin a new version of the patch set, with the last patch squashed into the third one and with the byte-wide atomic flag replacing the bitset_t?

I think the implementation using the atomics may end up doing the same assumptions as the approach I proposed but it would be easier to discuss with a concrete code at hand (= I will not quite yet stop suggesting an approach with minimal atomic and synchronization operations).

Reply:
Initially I kept this patch separate, as a way to demonstrate the change. Squashed now into third patch. Is a general uneasiness about cache propagation of flag clear in `alloc_session` and flag set in `crypto_init`. E.g. set-clear-set done in fast enough way. Maybe I'm wrong and flags will propagate in such case.

Reply:
Maybe @WonderfulVoid can comment on this issue.